### ISSUE TITLE
Analytics Implementation or QA Support for Caregivers

### ISSUE THREAD

**Issue by ala-yna** [department-of-veterans-affairs/va.gov-team#8081]
# Analytics Implementation or QA Request

## What this form is for
Use this template to request Google Tag Manager and Google Analytics implementation or QA for your product. Please fill out all brackets. 

> _Please see the following information for implementation timeline:_
> New Teams: Implementation will usually be completed in 4 sprints:
> 
> - Sprint 1: Discovery/Orientation
> - Sprint 2: FE Implementation
> - Sprint 3: Tagging Implementation
> - Sprint 4: QA
> 
> All other teams: Implementation will usually be completed in 2 sprints
> 
> - Sprint 1: Discovery and FE Implementation
> - Sprint 2: Tagging Implementation and QA

## Description
- Is this request for **Google Analytics** implementation or QA?

- Who is this request for?

    - Caregivers Team (@ala-yna, @LinnJS, @ccgongie) #vsa-caregivers in DSVA Slack

    _Provide any relevant contact information and link this issue to the insights team on the DSVA Slack Channel: `#VFS-Platform-Support`._
    
- What team/project is this work for?
 - Caregivers, online 10-10CG dark release
   
    
- Provide the Domain where Google Analytics needs to be implemented
-  https://www.va.gov/health-care/family-caregiver-benefits/comprehensive-assistance/ (owned by Public Websites)
- Plus form intro page (https://www.va.gov/family-member-benefits/apply-for-caregiver-program-form-10-10cg/)
   
- Please provide any upcoming deadlines or important milestones
    - Dark Release on May 29th

- Typical Website Interactions to Track

    - *Conversions - Understanding the User Flow:* _Provide URLs that are part of the application from introduction page to successful form completion; conditional pages that are visible depending on an applicant's answer should be marker accordingly._
    
    - Program page --> Form intro page
    - Form intro page --> Form start

    
    - *Other User Interactions*: _Provide details on other interactions and metrics, like click to help text links, that are helpful in understanding user behavior._
    
    - Click to learn more
    - Time spent completing the form (if possible)
    - Veteran could not be verified

- Sample Users when authentication is required _**do not put staging credentials in your `va.gov-team` ticket; store or reference them in a .md file in the `va.gov-team-sensitive` repository**_

    - Form in authenticated but required valid veteran SSN

- Test Users to replicate the event _**do not put staging credentials in your `va.gov-team` ticket; store or reference them in a .md file in the `va.gov-team-sensitive` repository**_

    - N/A

- Certain conditions/responses that need to be reviewed

    - Veteran cannot be verified

- Additional Comments

    - [Add any comments or questions for the Insights team here (if applicable).]

> Please leave the following blank

## Acceptance Criteria
- [ ] QA has been completed
- [ ] VFS team completes brief [VSP Collaboration Cycle Feedback](https://adhoc.optimalworkshop.com/questions/20260uu8-0-0/questions/before) survey

## Definition of Done
- [ ] All appropriate issue tagging is completed
- [ ] All AC completed

**Comment by jonwehausen** [comment]
@ala-yna Thank you! Acknowledging receipt, will be in touch soon.

**Comment by jonwehausen** [comment]
Hi @ala-yna ,

When you have a moment, we would like to schedule a kickoff meeting with you to learn more about your tracking wants / needs. Feel free to reach out here or over Slack with your availability for the upcoming week and we can get something on the calendar.

**Comment by jonwehausen** [comment]
## Datalayer Requirements
cc: @ala-yna @LinnJS 

## Submission Events
Description of Interaction | dataLayer Example
------------ | ------------- 
_Start form 10-10CG_ | `dataLayer.push({`<br>`'event': 'caregivers-10-10cg-start-form'`<br>`});`
_All submissions of form 10-10-CG_ | `dataLayer.push({`<br>`'event': 'caregivers-10-10cg-submission'`<br>`});`
_**Successful** submissions of form 10-10cg_ | `dataLayer.push({`<br>`'event': 'caregivers-10-10cg-submission-successful'`<br>`});`
_**Failed** submissions of form 10-10cg | `dataLayer.push({`<br>`'event': 'caregivers-10-10cg-submission-failed'`<br>`});`
_Veteran could not be verified_ | `dataLayer.push({`<br>`'event': 'caregivers-error',`<br>`error-key': 'XXX_veteran_could_not_be_verified'`<br>`});` <br> ***replace xxx with relevant error code***

@ala-yna @LinnJS when you and your team is ready -- we have the DL events drafted ready for your implementation.

**Comment by jonwehausen** [comment]
Hi @LinnJS ,

Just wanted to let you know the DL requirements above are ready for implementation when you and youre team are ready. Thanks in advance!

**Comment by joanneesteban** [comment]
@ala-yna - is this on the queue for Caregivers? We're ready to implementation when you're ready!

**Comment by ala-yna** [comment]
Yes! A little backlogged at the moment. The ticket we're using to track is: https://app.zenhub.com/workspaces/vft-59c95ae5fda7577a9b3184f8/issues/department-of-veterans-affairs/va.gov-team/7849 :)

**Comment by jonwehausen** [comment]
Ticket from FE: https://github.com/department-of-veterans-affairs/va.gov-team/issues/3825

**Comment by LinnJS** [comment]
@jonwehausen going to make another ticket for event "Veteran could not be verified" due to the FE not receiving an error code specific to MPI/MVI due to security reasons. @kevinmirc will have to implement this tracking on the BE. Is there documentation on how this is done on vets-api?

**Comment by jonwehausen** [comment]
Hi @LinnJS , 

Thank you for bringing this to our attention. There aren't too many backend events we pull in to Google Analytics other than what comes OOB. If there's no error code returned, that's totally fine, we can just pass in `veteran_could_not_be_verified` without the error code up front.

**Comment by jonwehausen** [comment]
### QA of DataLayer
Description of Interaction | DL Screenshot | Test Status 
------------ | ------------- | -------------
Start form | ![image](https://user-images.githubusercontent.com/48527022/87563116-df29c880-c68c-11ea-83b3-85aaf89522f4.png) | ✔️ **PASS**
Submission event | ![image](https://user-images.githubusercontent.com/48527022/87565123-3a5cba80-c68f-11ea-8e89-1648a2460a8d.png)  | ✔️ **PASS**
Submission failed | ![image](https://user-images.githubusercontent.com/48527022/87565015-19946500-c68f-11ea-81a5-9e51a4a4d28b.png) | ✔️ **PASS**

@LinnJS Should we be able to test the `--submission-successful` event at this point?

**Comment by joanneesteban** [comment]
@ala-yna @LinnJS 

Re-upping the above comment^

Can we get a test user to test the submission successful event? We'll need this before we can track submissions on the form.

Thanks!

**Comment by ala-yna** [comment]
Hi @joanneesteban and @jonwehausen - Use any of the users [here](https://github.com/department-of-veterans-affairs/va.gov-team-sensitive/blob/master/Administrative/vagov-users/mvi-staging-users.csv) for the Veteran's information and you should get a successful submission. Thanks!

**Comment by bmcgrady-ep** [comment]
### QA of DataLayer
Description of Interaction | DL Screenshot | Test Status 
------------ | ------------- | -------------
Submission Successful | 
![image](https://user-images.githubusercontent.com/66263904/88429851-3fff8200-cdac-11ea-8543-152dd67d9e50.png)
 | ✔️ PASS

QA looks good for the submission successful event. The new staging URL for the form is https://staging.va.gov/family-member-benefits/apply-for-caregiver-assistance-form-10-10cg/introduction.

@ala-yna @LinnJS 
The Submission Successful event passed QA and is ready to be deployed. The sooner it is deployed, the sooner we will start receiving data for the KPI Dashboard.

**Comment by bmcgrady-ep** [comment]
This is available to be deployed with the daily 3:00 deployment.

**Comment by jonwehausen** [comment]
Functionality is deployed -- will keep dashboard ticket open following launch

**Comment by jonwehausen** [comment]
### Reopening in regards to caregivers error tracking


Speaking with @kevinmirc via Slack, the tracking of caregivers errors must be done from the backend rather than the frontend dataLayer due to the failure experience not changing for the user and metadata towards the error not exposed to the front end. For this reason, we are needing to construct the hit payload from the BE to send these events to Google Analytics.

### Here is a template example of what the full tracking payload will need to look like: 
```html
v=1&t=event&tid=UA-CHANGE-ME&cid=CHANGE_ME&ec=API%20Calls&ea=Forms%20-%20Family%20Member%20Benefits&el=caregivers-error&cd30=Modernized&cd57=XXX_CHANGE_ME_QUALIFICATION_DESCRIPTION&cg1=Family%20Member%20Benefits&cg3=Family%20Member%20Benefits&gtm=2wg8c0WFJWBD&dl=https%3A%2F%2FCHANGE_ME.va.gov%2Ffamily-member-benefits%2Fapply-for-caregiver-assistance-form-10-10cg%2Freview-and-submit
```

Payload Variable | Remain Consistent | Sample Value
---------- | ----------- | -------------
`&v1` | ✔️ | `1`
`&t` | ✔️  | `event`
`&tid` | ❌ | `UA-50123418-17` **Change according to environment, dev/staging is GA property `UA-50123418-17` while production is `UA-50123418-16`**
`&cid` | ❌ | `1864532164.1562686258` **Change according to the GA Client ID of the user taken from the browser session**
`&cg1` | ✔️ | `Family Member Benefits`
`&cg3` | ✔️ | `Family Member Benefits`
`&cd30` | ✔️ | `Remain consistent as Modernized`
`&ec` | ✔️ | `API Calls`
`&ea` | ✔️ | `Forms - Family Member Benefits`
`&el` | ✔️ | `caregivers-error`
`&cd57` | ❌  | `XXX_DESCRIPTION_OF_ERROR`
`&gtm` | ✔️ | `2wg8c0WFJWBD`
`&dl` | ❌  | `https://staging.va.gov/family-member-benefits/apply-for-caregiver-assistance-form-10-10cg/review-and-submit`

@kevinmirc Please let us know any further questions you have around this. We will want to test this in staging first for each of the errors you're looking to track before proceeding. Thanks again for your efforts here!

cc: @bsmartin-ep @bmcgrady-ep for visibility

**Comment by kevinmirc** [comment]
@jonwehausen Thank you very much for collecting the parameters required to fire off this event from the API. I'll let you know when it's in staging so we can do a test.

**Comment by jonwehausen** [comment]
Met with @kevinmirc this morning to discuss debugging with the user explorer in GA and determined next steps in the implementation / QA of tracking backend errors as described [here](https://github.com/department-of-veterans-affairs/va.gov-team/issues/8081#issuecomment-677682631).

### Next Steps
- @kevinmirc will notify us when we has the error hits configured within staging. We can then work with him to fire off test events pointed at our staging property `UA-50123418-17` 
- From there, we will validate and QA readiness for production

**Comment by lihanli** [comment]
@jonwehausen i don't see the `gtm` parameter listed in [google's api docs](https://developers.google.com/analytics/devguides/collection/protocol/v1/parameters), are we sure their api can understand it?

**Comment by jonwehausen** [comment]
Hi @lihanli , 

Thanks for the call-out on the `gtm` parameter....The `gtm` parameter should work -- however, we may be able to leave out due the fact we're sending the event hit directly here

We were hoping as part of this QA we could do a live fire of this event from the backend and we can validate in real-time that we received in Google Analytics.

Thanks and look forward to hearing from you. 

Best,
Jon Wehausen

**Comment by lihanli** [comment]
@jonwehausen do you want to test it from the console in the backend or wait until it's implemented in the codebase to test it when there is a form submission failure?

**Comment by lihanli** [comment]
@jonwehausen is the user's GA Client ID always going to be available? what if they are using adblock?

**Comment by jonwehausen** [comment]
@lihanli Great questions: 

1) We can wait until implemented in the codebase and test along with a form submission failure
2) In the cases where it's not available, the hit won't be registered to GA anyhow.

**Comment by jonwehausen** [comment]
Spoke with @kevinmirc and @lihanli this afternoon -- we believe it's best here to leave this backend event out of GA at this time, as this would need to expand to other backend tagging work for successful and failed submissions that is already best served in other systems. 

@kevinmirc we will close this when we have confirmed with you and your team we will leave this implementation out. 

Thanks again

**Comment by kevinmirc** [comment]
@lihanli we can move forward with reverting "caregivers google analytics event" on vets-api. Thank you!
